src/huggingface_hub/utils/_deprecation.py

```diff
@@ -131,3 +131,100 @@ def inner_f(*args, **kwargs):
         return inner_f
 
     return _inner_deprecate_method
+
+
+def _deprecate_list_output(*, version: str):
+    """Decorator to deprecate the usage as a list of the output of a method.
+
+    To be used when a method currently returns a list of objects but is planned to return
+    an generator instead in the future. Output is still a list but tweaked to issue a
+    warning message when it is specifically used as a list (e.g. get/set/del item, get
+    length,...).
+
+    Args:
+        version (`str`):
+            The version when output will start to be an generator.
+    """
+
+    def _inner_deprecate_method(f):
+        @wraps(f)
+        def inner_f(*args, **kwargs):
+            list_value = f(*args, **kwargs)
+            return DeprecatedList(
+                list_value,
+                warning_message=(
+                    "'{f.__name__}' currently returns a list of objects but is planned"
+                    " to be a generator starting from version {version} in order to"
+                    " implement pagination. Please avoid to use"
+                    " `{f.__name__}(...).{attr_name}` or explicitly convert the output"
+                    " to a list first with `list(iter({f.__name__})(...))`.".format(
+                        f=f,
+                        version=version,
+                        # Dumb but working workaround to render `attr_name` later
+                        # Taken from https://stackoverflow.com/a/35300723
+                        attr_name="{attr_name}",
+                    )
+                ),
+            )
+
+        return inner_f
+
+    return _inner_deprecate_method
+
+
+def _empty_gen() -> Generator:
+    # Create an empty generator
+    # Taken from https://stackoverflow.com/a/13243870
+    return
+    yield
+
+
+# Build the set of attributes that are specific to a List object (and will be deprecated)
+_LIST_ONLY_ATTRS = frozenset(set(dir([])) - set(dir(_empty_gen())))
+
+
+class DeprecateListMetaclass(type):
+    """Metaclass that overwrites all list-only methods, including magic ones."""
+
+    def __new__(cls, clsname, bases, attrs):
+        # Check consistency
+        assert (
+            "_deprecate" in attrs
+        ), "A `_deprecate` method must be implemented to use `DeprecateListMetaclass`."
+        assert (
+            list in bases
+        ), "Class must inherit from `list` to use `DeprecateListMetaclass`."
```

Comment:
In general, I would favor raising errors with the appropriate types rather than `assert`, but if you feel like this is more coherent this way we can leave it be.

Reply:
Yep true. I usually use assertions only for internal logic issues (like here where the metaclass is really not meant to be reused by anyone). But agree on changing this :)